common/unsafe/src/main/java/org/apache/spark/sql/catalyst/util/CollationFactory.java

```diff
@@ -117,76 +119,445 @@ public Collation(
+        collationId ^= accentSensitivityOrdinal << ACCENT_SENSITIVITY_OFFSET;
+        collationId ^= caseConversionOrdinal << CASE_CONVERSION_OFFSET;
+        collationId ^= localeOrdinal << LOCALE_OFFSET;
+        if (collationId != 0 || caseConversionOrdinal >= CaseConversion.values().length ||
+            localeOrdinal >= ICULocaleNames.length) {
+          throw SparkException.internalError("Invalid ICU collation id " + originalCollationId);
+        } else {
+          CaseSensitivity caseSensitivity = CaseSensitivity.values()[caseSensitivityOrdinal];
+          AccentSensitivity accentSensitivity =
+            AccentSensitivity.values()[accentSensitivityOrdinal];
+          CaseConversion caseConversion = CaseConversion.values()[caseConversionOrdinal];
+          String locale = ICULocaleNames[localeOrdinal];
+          return new CollationSpecICU(locale, caseSensitivity, accentSensitivity, caseConversion);
+        }
+      }
+
+      @Override
+      protected int getCollationId() {
+        int collationId = 0;
+        collationId |= ImplementationProvider.ICU.ordinal() << IMPLEMENTATION_PROVIDER_OFFSET;
+        collationId |= caseSensitivity.ordinal() << CASE_SENSITIVITY_OFFSET;
+        collationId |= accentSensitivity.ordinal() << ACCENT_SENSITIVITY_OFFSET;
+        collationId |= caseConversion.ordinal() << CASE_CONVERSION_OFFSET;
+        collationId |= ICULocaleToId.get(locale);
+        return collationId;
+      }
+
+      @Override
+      protected Collation buildCollation() {
+        ULocale.Builder builder = new ULocale.Builder();
+        builder.setLocale(ICULocaleMap.get(locale));
+        if (caseSensitivity == CaseSensitivity.CS &&
+            accentSensitivity == AccentSensitivity.AS) {
+          builder.setUnicodeLocaleKeyword("ks", "level3");
+        } else if (caseSensitivity == CaseSensitivity.CS &&
+            accentSensitivity == AccentSensitivity.AI) {
+          builder
+            .setUnicodeLocaleKeyword("ks", "level1")
+            .setUnicodeLocaleKeyword("kc", "true");
+        } else if (caseSensitivity == CaseSensitivity.CI &&
+            accentSensitivity == AccentSensitivity.AS) {
+          builder.setUnicodeLocaleKeyword("ks", "level2");
+        } else if (caseSensitivity == CaseSensitivity.CI &&
+            accentSensitivity == AccentSensitivity.AI) {
+          builder.setUnicodeLocaleKeyword("ks", "level1");
+        }
+        ULocale resultLocale = builder.build();
+        Collator collator = Collator.getInstance(resultLocale);
+        Comparator<UTF8String> comparator = (s1, s2) -> collator.compare(
+          caseConversion(resultLocale, s1), caseConversion(resultLocale, s2));
+        collator.freeze();
+        return new Collation(
+          collationName(),
+          collator,
+          comparator,
+          ICUCollatorVersion,
+          s -> (long) collator.getCollationKey(s.toString()).hashCode(),
+          collationId == UNICODE_COLLATION_ID,
```

Comment:
I think other collations can `supportBinaryEquality` too?

for example: `SRB_CS_AS` is essentially binary

Reply:
I'm not sure it's safe to go with this assumption in general. `CS` and `AS` combination maps to tertiary level, but there is an additional "identity" level, for which ICU [docs](https://unicode-org.github.io/icu/userguide/collation/concepts.html) specifies:
> For example, Hebrew cantillation marks are only distinguished at this level.

This is an example where strings could compare equal at tertiary level (`CS` and `AS`) but would differ in their binary representation. We shouldn't make any assumptions for a particular language.

Reply:
Ok, let's go them being non-binary. In future, we may want to add _BIN suffix that will indicate that we are doing binary comparisons (which is going to be faster).